src/transformers/generation_beam_constraints.py

```diff
@@ -202,6 +211,170 @@ def copy(self, stateful=False):
+            start = start[current_token]
+
+        next_tokens = list(start.keys())
+
+        return next_tokens
+
+    def reached_leaf(self, current_seq):
+        next_tokens = self.next_tokens(current_seq)
+
+        return -1 in next_tokens
+
+
+class DisjunctiveConstraint(Constraint):
+    r"""
+    A special [`Constraint`] that is fulfilled by fulfilling just one of several constraints.
+
+    Args:
+        constraints (`List[Constraint]`):
+            The list of [`Constraint`] objects, fulfillment of just one of which fulfills this
+            [`DisjunctiveConstraint`].
+    """
+
+    def __init__(self, nested_token_ids: List[Constraint]):
+        super(Constraint, self).__init__()
+
+        self.token_ids = nested_token_ids
+        self._check_valid(self.token_ids)
+
+        self.trie = DisjunctiveTrie([ids.tolist() for ids in self.token_ids])
+
+        self.seqlen = self.trie.max_height
+        self.current_seq = []
+        self.completed = False
+
+    def _check_valid(self, nested_token_ids):
+        """
+        Checks whether nested_token_ids is valid.
+        1. Checks Type
+        2. Checks for illegal subsets
+            We're checking if there is a `seq_k` in `nested_token_ids` such that `seq_k` is 100% a subset found in some
+            seq in `nested_token_ids`. We can't allow this because it leads to a perverse interpretation of
+            "completeness". Luckily, we only need to check the beginning of the sequence since it's a trie
+            representation.
+        """
+        if not isinstance(nested_token_ids, list):
+            nested_token_ids = nested_token_ids.tolist()
+            self.token_ids = nested_token_ids
+
+        for k, seq_k in enumerate(nested_token_ids):
+            is_tensor = isinstance(seq_k, torch.Tensor)
+            is_int_list = isinstance(seq_k, List) and isinstance(seq_k[0], int)
+            is_int_tensor = is_single_int_tensor(seq_k)
+
+            not_positive = torch.any(seq_k < 0) if is_tensor else len([t for t in seq_k if t < 0]) > 0
+            if isinstance(seq_k, int) or not (is_int_list or is_int_tensor) or not_positive:
+                raise ValueError(
+                    f"Sequences within `nested_token_ids` has to be a single list or tensor of positive integers but is {seq_k}"
+                )
+
+            seq_k = seq_k.tolist() if is_int_tensor else seq_k
+            for i, seq_i in enumerate(nested_token_ids[k + 1 :]):
+                seq_i = seq_i.tolist() if is_single_int_tensor(seq_i) else seq_i
+
+                if seq_k == seq_i[: len(seq_k)] or seq_k[: len(seq_i)] == seq_i:
+                    raise ValueError(
+                        "No sequence in `nested_token_ids` can be a complete subset of another, but found"
+                        f" sequences {seq_k} and {seq_i}."
+                    )
+
+            self.token_ids[k] = torch.tensor(seq_k)
```

Comment:
This code is extremely hard to read and most likely contains bugs (just looking at the amount of casting to and from tensors)

If we want to check if there's a subset of  a sequence in another, we don't even need to do that `O(n²)` comparisons, just create the Trie, and raise and exception if a node is a not a pure leaf (if it has other children). That should I think cover what you are trying to do here. Checking that directly in the trie will be `O(n)`.

 Also this `is_int_list = isinstance(seq_k, List) and isinstance(seq_k[0], int)` is most likely incorrect since users can pass a list that is `[24, MyCustomObject()]`. Which is why IMO, instead of trying to figure what object was given to you, you should expect it to be of some form and run with it. And fix every caller of your function such that they give you a consistent input.

Forcing `token_ids` to be `List[List[int]]` seems to me like the best thing (but I have no idea if this is possible/relevant when running those constraints on a GPU generation for instance. The next best thing is a 2-Tensor (but you loose the option of having different lengths of token_ids.

Reply:
Agreed! This code is completely gone now and everything is overhauled to force `token_ids` to be `List[List[int]]` and will raise a ValueError otherwise.